firebase-common/src/main/java/com/google/firebase/heartbeatinfo/HeartBeatInfoStorage.java

```diff
@@ -30,17 +36,38 @@
 
   private static final String preferencesName = "FirebaseAppHeartBeat";
 
+  // As soon as you hit the limit of heartbeats. The number of stored heartbeats is halved.
+  private static final int HEART_BEAT_COUNT_LIMIT = 200;
+
+  private static final SimpleDateFormat FORMATTER = new SimpleDateFormat("dd/MM/yyyy z");
+
+  // Stores a key value mapping from timestamp to the sdkName and heartBeat code.
+  private static final String storagePreferencesName = "FirebaseAppHeartBeatStorage";
+
   private final SharedPreferences sharedPreferences;
+  private final SharedPreferences heartBeatSharedPreferences;
+  private int heartBeatCount;
 
   private HeartBeatInfoStorage(Context applicationContext) {
     this.sharedPreferences =
         applicationContext.getSharedPreferences(preferencesName, Context.MODE_PRIVATE);
+    this.heartBeatSharedPreferences =
+        applicationContext.getSharedPreferences(preferencesName, Context.MODE_PRIVATE);
+    this.heartBeatCount = 0;
   }
 
   @VisibleForTesting
   @RestrictTo(RestrictTo.Scope.TESTS)
-  HeartBeatInfoStorage(SharedPreferences preferences) {
+  HeartBeatInfoStorage(
+      SharedPreferences preferences, SharedPreferences heartBeatSharedPreferences) {
     this.sharedPreferences = preferences;
+    this.heartBeatSharedPreferences = heartBeatSharedPreferences;
+  }
+
+  @VisibleForTesting
+  @RestrictTo(RestrictTo.Scope.TESTS)
+  int getHeartBeatCount() {
+    return this.heartBeatCount;
```

Comment:
looks like you're updating it inside synchronized, but reading it unprotected, this is a race. consider making this method synchronized as well

Reply:
hm, actually: are you using this variable to control when to clean up stale heartbeats? if so, is it ok that this variable will not survive app restarts? wouldn't it cause storage to grow in an unbounded manner?

Reply:
Hmm the method is just restricted for Testing so hadn't made it synchronized.  Hmmm but good catch, maybe it makes sense to put this as well under shared preferences?